experimental/Schemes/AffineSchemes.jl

```diff
@@ -455,34 +784,59 @@ function preimage(
   return Spec(MPolyQuoLocalizedRing(R, ideal(R, new_gens) + modulus(OO(X)), inverted_set(OO(X))*new_units))
 end
 
-function is_isomorphism(f::SpecMor)
+function preimage(f::AbsSpecMor, Z::AbsSpec{<:Ring, <:MPolyRing}; check::Bool=true)
```

Comment:
I do not see Z used in the function anywhere. Is that, because Z is automatically the whole codomain here or do we at least need to check compatibility with the codomain of f (as a sanity check)?

Reply:
Ok, this is a mistake.

Reply:
Looking at it again: It's not a mistake. Since `Z = Spec(P)` for a polynomial ring, it must necessarily be the full ambient space of the codomain of `f`, so the preimage is simply a copy of the domain.

Reply:
Let me put the question more directly: How do you make sure that the user has not provided some Z which has nothing to do with f at all (maybe by mistake or typo in the name or because a previous computation had an unexpected result)? As you are not touching Z anywhere, no sanity check (apart from the fact that Z is a Spec of some polynomial ring) is performed -- neither explicitly nor implicitly.